internal/pkg/git/git.go

```diff
@@ -116,56 +118,69 @@ func (r *Repository) Clear() {
 	r.fs = nil
 }
 
-func (r *Repository) CommitHash() (string, error) {
+func (r *Repository) CommitHash(ctx context.Context) (string, error) {
 	r.lock.RLock()
 	defer r.lock.RUnlock()
-	result, err := r.runGitCmd("rev-parse", "HEAD")
+	result, err := r.runGitCmd(ctx, "rev-parse", "HEAD")
 	if err != nil {
 		return "", utils.PrefixError("cannot get repository hash", fmt.Errorf(result.stdErr))
 	}
 	return strings.TrimSuffix(result.stdOut, "\n"), nil
 }
 
 // Load a path from the remote git repository, if sparse mode is used.
-func (r *Repository) Load(path string) error {
+func (r *Repository) Load(ctx context.Context, path string) error {
 	if !r.sparse {
 		return fmt.Errorf("sparse checkout is not allowed")
 	}
-	if _, err := r.runGitCmd("sparse-checkout", "add", fmt.Sprintf("/%s", path)); err != nil {
+	if _, err := r.runGitCmd(ctx, "sparse-checkout", "add", fmt.Sprintf("/%s", path)); err != nil {
 		return err
 	}
-	if _, err := r.runGitCmd("checkout"); err != nil {
+	if _, err := r.runGitCmd(ctx, "checkout"); err != nil {
 		return err
 	}
 	return nil
 }
 
-func (r *Repository) Pull() error {
+func (r *Repository) Pull(ctx context.Context) error {
+	// Check remote changes
+	if err := r.fetch(ctx); err != nil {
+		return err
+	}
+
+	// Acquire write lock, from the repository must not be read during this time
 	r.lock.Lock()
 	defer r.lock.Unlock()
 
-	// Check remote changes
-	result, err := r.runGitCmd("fetch", "origin")
+	// Reset is used, because it works also with force push (edge-case)
+	result, err := r.runGitCmd(ctx, "reset", "--hard", fmt.Sprintf("origin/%s", r.ref))
 	if err != nil {
-		return utils.PrefixError("cannot fetch repository", fmt.Errorf(result.stdErr))
+		return utils.PrefixError("cannot reset repository to the origin", fmt.Errorf(result.stdErr))
 	}
```

Comment:
2. Na fetch som vytvoril samostatny zamok, ... fetch trva dlho, kedze je to sietova oparacia, ale nemeni FS, takze takto nebude blokovat requesty.